File: Problem144_clone-graph.java
// Time Complexity: O(v + e)
// Space Complexity: O(v)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach

class Solution {
    HashMap<Node, Node> map;
    public Node cloneGraph(Node node) {
        map = new HashMap<>();
        if(node == null)
            return node;
        Queue<Node> q = new LinkedList<>();
        Node copyNode = clone(node);
        q.add(node);
        while(!q.isEmpty()) {
            Node curr = q.poll();
            List<Node> children = curr.neighbors;
            for(Node child: children) {
                if(!map.containsKey(child)) {
                    clone(child);
                    q.add(child);
                }
            // create the edges of the current Node copy
            // get the deep copy of current
            // add to that deep copy the deep copy of child of curr
            map.get(curr).neighbors.add(map.get(child));
            }
        }
        // return map.get(node);
        return copyNode;
    }
    private Node clone(Node node) {
        if(map.containsKey(node)) return map.get(node);
        Node copy = new Node(node.val);
        map.put(node, copy);
        return copy;
    }
}

File: Problem-144 Clone Graph.py
# 133. Clone Graph
# https://leetcode.com/problems/clone-graph/

# Logic: We make a hashmap with original node as key and a copy node as value. 
# We will use the hashmap as a visited set also. Start from node given and 
# create its children if copy is not created. Only visit nodes who's copy 
# has not been cretaed. When the copy is created only then we add it to 
# queue as it has not been visted so far.

# Time Complexity: O(V+E)
# Space Complexity: O(V)

"""
# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""
from collections import deque
class Solution:
    def cloneGraph(self, node: 'Node') -> 'Node':
        if node is None:
            return None
        
        hashmap = dict()
        hashmap[node] = Node(node.val)
        
        q = deque()
        q.append(node)
        
        while q:
            curNode = q.popleft()
            neigh = curNode.neighbors
            
            for i in neigh:
                if i not in hashmap:
                    hashmap[i] = Node(i.val)
                    q.append(i)
                hashmap[curNode].neighbors.append(hashmap[i])
        
        return hashmap[node]